File: src/python_scripts/cgalpy_examples/Arrangement_on_surface_2/aos2_visual_buffer_summary.cpp
#include "arr_exact_construction_segments.h"

#include <CGAL/draw_arrangement_2.h>
#include <CGAL/Graphics_scene.h>

#include <iomanip>
#include <iostream>
#include <numeric>

namespace {

const char* py_bool(bool value) { return value ? "True" : "False"; }

void print_buffer_summary(const char* name,
                          CGAL::Graphics_scene::Buffers index,
                          const CGAL::Graphics_scene& scene) {
  const auto buffer = scene.get_array_of_index(index);
  const double sum = std::accumulate(buffer.begin(), buffer.end(), 0.0);

  std::cout << name
            << " elements " << scene.number_of_elements(index)
            << " size " << scene.get_size_of_index(index)
            << " floats " << buffer.size()
            << " sum " << std::fixed << std::setprecision(6) << sum
            << '\n';
}

} // namespace

int main() {
  Segment segments[] = {
    Segment(Point(1, 0), Point(2, 4)),
    Segment(Point(5, 0), Point(5, 5)),
    Segment(Point(1, 0), Point(5, 3)),
    Segment(Point(0, 2), Point(6, 0)),
    Segment(Point(3, 0), Point(5, 5))
  };

  Arrangement arr;
  insert(arr, segments, segments + sizeof(segments) / sizeof(Segment));

  CGAL::Graphics_scene scene;
  std::cout << "empty_before " << py_bool(scene.empty()) << '\n';

  CGAL::add_to_graphics_scene(arr, scene);

  std::cout << "empty_after " << py_bool(scene.empty()) << '\n';
  std::cout << "is_two_dimensional " << py_bool(scene.is_two_dimensional()) << '\n';

  // Only geometry buffers are compared. Color buffers can depend on
  // random/default coloring and are intentionally excluded.
  print_buffer_summary("POS_POINTS", CGAL::Graphics_scene::POS_POINTS, scene);
  print_buffer_summary("POS_SEGMENTS", CGAL::Graphics_scene::POS_SEGMENTS, scene);
  print_buffer_summary("POS_RAYS", CGAL::Graphics_scene::POS_RAYS, scene);
  print_buffer_summary("POS_LINES", CGAL::Graphics_scene::POS_LINES, scene);
  print_buffer_summary("POS_FACES", CGAL::Graphics_scene::POS_FACES, scene);
  print_buffer_summary("SMOOTH_NORMAL_FACES", CGAL::Graphics_scene::SMOOTH_NORMAL_FACES, scene);
  print_buffer_summary("FLAT_NORMAL_FACES", CGAL::Graphics_scene::FLAT_NORMAL_FACES, scene);

  return 0;
}

File: src/python_examples/cgalpy_examples/Arrangement_on_surface_2/aos2_visual_buffer_summary.py
#!/usr/bin/python

import importlib
import sys

if len(sys.argv) < 2:
  lib = "CGALPY"
else:
  lib = sys.argv[1]

CGALPY = importlib.import_module(lib)
Aos2 = CGALPY.Aos2
Bvr = CGALPY.Bvr

Arrangement = Aos2.Arrangement_2
Point = Arrangement.Geometry_traits_2.Point_2
Segment = Arrangement.Geometry_traits_2.X_monotone_curve_2

segments = [
  Segment(Point(1, 0), Point(2, 4)),
  Segment(Point(5, 0), Point(5, 5)),
  Segment(Point(1, 0), Point(5, 3)),
  Segment(Point(0, 2), Point(6, 0)),
  Segment(Point(3, 0), Point(5, 5)),
]

arr = Arrangement()
Aos2.insert(arr, segments)

scene = Bvr.Graphics_scene()
print("empty_before", scene.empty())
Bvr.add_to_graphics_scene(arr, scene)
print("empty_after", scene.empty())
print("is_two_dimensional", scene.is_two_dimensional())

# Only geometry buffers are compared. COLOR_FACES was observed to be
# nondeterministic across repeated C++/Python runs.
indices = [
  ("POS_POINTS", Bvr.POS_POINTS),
  ("POS_SEGMENTS", Bvr.POS_SEGMENTS),
  ("POS_RAYS", Bvr.POS_RAYS),
  ("POS_LINES", Bvr.POS_LINES),
  ("POS_FACES", Bvr.POS_FACES),
  ("SMOOTH_NORMAL_FACES", Bvr.SMOOTH_NORMAL_FACES),
  ("FLAT_NORMAL_FACES", Bvr.FLAT_NORMAL_FACES),
]

for name, index in indices:
  buf = scene.buffer(index)
  print(
    name,
    "elements", scene.number_of_elements(index),
    "size", scene.get_size_of_index(index),
    "floats", len(buf),
    "sum", format(sum(buf), ".6f"),
  )
